src/process/extract/Extractor.h

```diff
@@ -0,0 +1,174 @@
+     * @param mode              Extraction mode, currently either lines or segment
+     * @param start             Start extraction from this line or extract this segment
+     * @param count             Extract a maximum of lineCount lines OR define the number of total segments
+     * @param enableDebugging   Used for interactive debugging and unit tests
+     */
+    explicit Extractor(const shared_ptr<Source> &fastqfile,
+                       const shared_ptr<Source> &indexFile,
+                       const shared_ptr<Sink> &resultSink,
+                       bool forceOverwrite,
+                       ExtractMode mode, u_int64_t start, u_int64_t count, uint extractionMulitplier,
+                       bool enableDebugging);
+
+    virtual ~Extractor();
+
+    ExtractMode getExtractMode() { return mode; }
+
+    shared_ptr<Sink> getResultSink() { return resultSink; }
+
+    u_int64_t getStart() { return start; }
+
+    u_int64_t getCount() { return count; }
+
+    u_int64_t getStartingLine() { return startingLine; }
+
+    u_int64_t getLineCount() { return lineCount; }
+
+    /**
+     * Will call tryOpenAndReadHeader on the internal indexReader.
+     */
+    bool checkPremises();
+
+    /**
+     * For debugging and testing , will be overriden by extract(). Sets the initial amount of lines which will be
+     * omitted.
+     * TODO Move to test-aware subclass.
+     */
+    void setSkip(uint skip) {
+        this->skip = skip;
+    }
+
+    /**
+     * For debugging and testing , will be overriden by extract(). Sets the initial amount of lines which will be
+     * omitted.
+     * TODO Move to test-aware subclass.
+     */
+    void setFirstPass(bool firstPass) {
+        this->firstPass = firstPass;
+    }
+
+    /**
+     * Calculates the starting line and the line count based on the start, count and mode settings.
+     */
+    void calculateStartingLineAndLineCount();
+
+    /**
+     * For now directly to cout?
+     * @param start
+     * @param count
+     */
+    bool extract();
+
+    /**
+     * Process a decompressed chunk (NOT a complete decompressed block!) of data.
+     *
+     * Will reset the firstPass variable, if called for the first time.
+     *
+     * @param out The stream to put the data to
+     * @param str The string of decompressed chunk data
+     * @param startingIndexLine The index entry which was used to step into the gzip file.
+     * @return true, if something was written out or false otherwise.
+     */
+    bool processDecompressedChunkOfData(ostream *out, string str, const shared_ptr<IndexEntry> &startingIndexLine);
+
+    bool checkAndPrepareForNextConcatenatedPart(bool finalAbort);
+
+    void storeOrOutputLine(ostream *outStream, string line);
+
+    void storeLinesOfCurrentBlockForDebugMode();
+
+    /**
+     * Overriden to also pass through (copywise, safe but slow but also only with a few entries and in error cases)
```

Comment:
```suggestion
     * Overridden to also pass through (copywise, safe but slow but also only with a few entries and in error cases)
```
What does "copywise" mean?